### PR TITLE
Add foldMapK, reduceMapK

### LINKED ISSUE #2275: Add foldMapK and reduceMapK

### PR DESCRIPTION
**Pull request by kubukoz**
Closes #2275.

Not sure if we need to forward the target effect parameter of these functions in all the test classes, or whether we can just use `Id` or some other hardcoded type having a `MonoidK`, but for now I added appropriate implicits etc. everywhere.

### CODE REVIEW

**Comment by kubukoz** [comment]
BTW. we could also leave it as it is and use the `G` type instead of `Id` in the laws for `foldM`.

**Comment by kailuowang** [comment]
I am not sure about this one. It seems that a more generic approach would be provide an importable conversion from MonoidK to Monoid, something like 
```scala
implicit def monoidKAlgebra[G[_], A])(implicit ev: MonoidK[G]): Monoid[G[A]] = ev.algebra
```
and then wherever a `Monoid[G[A]]` is needed you can just import this method. I am not sure how I feel about having a bunch of pairs of methods that one takes a `Monoid` and the other takes a `MonoidK`.

**Comment by kubukoz** [comment]
I don't know about others, but I personally don't like implicit functions that aren't converting to extension classes (which is a workaround for no value classes in traits/classes)

**Comment by LukaJCB** [comment]
I think that particular method doesn't work when a type has different instances for Monoid and MonoidK. Maybe @tpolecat can chime in as he was the one with a use case for it :)

**Comment by kubukoz** [comment]
I just found another usecase for myself, WDYT about moving it to syntax for now (for bincompat) and continuing with it?

**Comment by LukaJCB** [comment]
Sounds good to me! :)

**Comment by ceedubs** [comment]
```scala
implicit def monoidKAlgebra[G[_], A])(implicit ev: MonoidK[G]): Monoid[G[A]] = ev.algebra
```

@kailuowang @kubukoz I don't think that we want to go down this path, as `MonoidK` and `Monoid` instances aren't always equivalent. For example in Cats the `Option` `Monoid` instance is based on combining with the inner `Semigroup` while the `MonoidK` instance is based on `orElse`. I think that there's a similar story for `Function1/Endo` but with function composition instead of `orElse`.

**Comment by kailuowang** [comment]
It's a fair point that MonoidK and Monoid instance could be different. Let's proceed with adding this to bincompat syntax traits.

**Comment by LukaJCB** [comment]
@kubukoz Are you still working on this? :)

**Comment by kubukoz** [comment]
tbh I completely forgot about it. I think I'll start over (especially with all these conflicts) and get something out there soon :)

**Comment by kubukoz** [comment]
@LukaJCB @kailuowang this one should be ready when the build is green :) sorry for the long wait

**Comment by kubukoz** [comment]
Rebased to resolve conflicts after #2670

File: core/src/main/scala/cats/syntax/reducible.scala [hunk 1]
```diff
@@ -9,3 +9,24 @@ trait ReducibleSyntax extends Reducible.ToReducibleOps {
...
+  implicit final def catsSyntaxReducibleOps0[F[_], A](fa: F[A]): ReducibleOps0[F, A] =
+    new ReducibleOps0[F, A](fa)
+}
+
+final class ReducibleOps0[F[_], A](val fa: F[A]) extends AnyVal {
+
+  /**
+   * Apply `f` to each element of `fa` and combine them using the
+   * given `SemigroupK[G]`.
+   *
```
Comment (LukaJCB) [inline core/src/main/scala/cats/syntax/reducible.scala]: I think in order for this to be valid you need to open it with `{{` and close after, that's why it's not actually running the doctests.

Comment (kubukoz) [reply core/src/main/scala/cats/syntax/reducible.scala]: True, plus there must be a space after each `scala>`. Fixed now :)

**Comment by kubukoz** [comment]
Looks about green :) @lukajcb

**Comment by kubukoz** [comment]
ping @kailuowang

**Pull request merged**